BUG: import of maybe_convert_indices in pandas.core.index.py, #10610

djgagne opened the pull request:
closes #10610 

I fixed the import statement and added a test to check for proper behavior when accessing a mixed-integer index with a list of values.

Commit:
Fixed import error in index.py and added exception raising for value checking in the case of a mixed-integer index. Added a test to verify in test_index

Commit:
BUG: GH10610 Fixed import error in index.py and added exception, test

Commit:
Fixed import error in index.py and added exception raising for value checking in the case of a mixed-integer index. Added a test to verify in test_index

Commit:
Merge remote-tracking branch 'origin/mixed_type_index' into mixed_type_index

Commit:
Merge remote-tracking branch 'origin/mixed_type_index' into mixed_type_index

# Conflicts:
#	pandas/tests/test_index.py

Commit:
Merge branch 'mixed_type_index' of https://github.com/djgagne/pandas into mixed_type_index

Commit:
Merge branch 'mixed_type_index' of https://github.com/djgagne/pandas into mixed_type_index

n test_index

jreback commented on pandas/core/index.py:
```diff
@@ -971,11 +971,13 @@ def _convert_list_indexer(self, keyarr, kind=None):
 
             if self.inferred_type == 'mixed-integer':
                 indexer = self.get_indexer(keyarr)
+                if not np.all(np.in1d(indexer, self.values)):
```
why is this line necessary?

djgagne replied:
Without the line, when passed a list containing index values, the function would output the data associated with the nearest integer index value. For other cases where an invalid index is passed to a DataFrame, an IndexError would be raised, so I made sure to explicitly check for that condition.

jreback replied:
your test doesn't hit that line though it goes thru
can u make a test that is explicitly caught there?

Commit:
BUG: #10610 Altered index checking to use existing checks to raise IndexError.

Commit:
Merge remote-tracking branch 'upstream/master' into mixed_type_index

djgagne commented:
I removed the index checking if-statement, and added a line that would change the -1 indices to equal the length of the Index being evaluated, so that the IndexError would be raised in maybe_convert_indices.

jreback commented on pandas/core/index.py:
```diff
@@ -973,9 +973,12 @@ def _convert_list_indexer(self, keyarr, kind=None):
+                # values outside the range of indices so as to trigger an IndexError in maybe_convert_indices
+                indexer[indexer < 0] = len(self)
+                from pandas.core.indexing import maybe_convert_indices
+                return maybe_convert_indices(indexer, len(self))
```
this is already done inside ``maybe_convert_indices`` why do you think its needed here?

djgagne replied:
It is necessary because negative indices passed into maybe_convert_indices are summed by the length of the array before the bounds checking is done. Because get_indexer assigns labels that are not found in the index with a value of -1, the last row or column is assigned to those positions instead of either NaNs or an exception being raised. With the line I added, the bounds checking in maybe_convert_indices will now throw an IndexError for out of bounds indices.

jreback commented:
ok, this is fine.

pls squash to a single commit. ping when green.

jreback commented on pandas/tests/test_index.py:
```diff
@@ -1729,6 +1729,11 @@ def test_equals_op_multiindex(self):
         tm.assert_numpy_array_equal(index_a == mi3, np.array([False, False, False]))
 
+    def test_multitype_list_index_access(self):
+        df = pd.DataFrame(np.random.random((10, 5)), columns=["a"] + [20, 21, 22, 23])
```
add the issue number as a comment here

jreback commented:
merged via https://github.com/pydata/pandas/commit/12df3ea39b28f752a0dcb43b87c4121387114cdf

thanks!
